lstchain/reco/dl1_to_dl2.py

```diff
@@ -51,6 +53,159 @@
+    # set to have equal weight in the interpolation)
+    w1 = np.where(np.isclose(closest_alt, second_closest_alt, atol=1e-4, rtol=0),
+                  0.5, (cos_tel_zd - c0) / (c1 - c0))
+    w0 = 1 - w1
+
+    # Update the dataframe:
+    dl2table = dl2table.assign(alt0=pd.Series(closest_alt).values,
+                               alt1=pd.Series(second_closest_alt).values,
+                               w0=pd.Series(w0).values,
+                               w1=pd.Series(w1).values)
+
+    return dl2table
+
+
+def predict_with_zd_interpolation(rf, param_array, features):
+    """
+    Obtain a RF prediction which takes into account the difference between
+    the telescope elevation (alt_tel, i.e. 90 deg - zenith) and those of the
+    MC training nodes. The dependence of image parameters (for a shower of
+    given characteristics) with zenith is strong at angles beyond ~50 deg,
+    due to the change in airmass. Given the way Random Forests work, if the
+    training is performed with a discrete distribution of pointings,
+    the prediction of the RF will be biased for pointings in between those
+    used in training. If zenith is used as one of the RF features, there will
+    be a sudden jump in the predictions halfway between the training nodes.
+
+    To solve this, we compute here two predictions for each event, one using
+    the elevation (alt_tel) of the training pointing which is closest to the
+    telescope pointing, and another one usimg the elevation of the
+    sceond-closest pointing. Then the values are interpolated (linearly in
+    cos(zenith)) to the actual zenith pointing (90 deg - alt_tel) of the event.
+
+    rf: sklearn.ensemble.RandomForestRegressor or RandomForestClassifier,
+    the random forest we want to apply (must contain alt_tel among the
+    training parameters)
+
+    param_array: pandas dataframe containing the features needed by theRF
+    It must also contain four additional columns: alt0, alt1, w0, w1, which
+    can be added with the function add_zd_interpolation_info. These are the
+    event-wise telescope elevations for the closest and 2nd-closest training
+    pointings (alt0 and alt1), and the event-wise weights (w0 and w1) which
+    must be applied to the RF prediction at the two pointings to obtain the
+    interpolated value at the actual telescope pointing. Since the weights
+    are the same (for a given event) for different RFs, it does not make
+    sense to compute them here - they are pre-calculated by
+    add_zd_interpolation_info
+
+    features: list of the names of the image features used by the RF
+
+    Return: interpolated RF predictions. 1D array for regressors (log energy,
+    or disp_norm), 2D (events, # of classes) for classifiers
+
+    """
+
+    # Type of RF (classifier or regressor):
+    is_classifier = isinstance(rf, RandomForestClassifier)
+
+    # keep original alt_tel values:
+    param_array.rename(columns={"alt_tel": "original_alt_tel"}, inplace=True)
```

Comment:
I think it would be easier to change the features rather than this game of name-swapping:
```
 features_copy =  features.copy()  # avoid modifying features outside the function
 alt_index_in_features = features_copy.index('alt_tel')
 features_copy[alt_index_in_features] = 'alt0'
 predict_0 = rf.predict(param_array[features_copy])
 features_copy[alt_index_in_features] = 'alt1'
 predict_1 = ...
```
that's it

Reply:
Much better indeed

Reply:
It seems that there is a check for the feature name, that is why I had gone for the column renaming (although I had forgotten by now)